esp/services/ws_dfu/ws_dfuService.cpp

```diff
@@ -6448,36 +6481,205 @@ bool CWsDfuEx::onDFUFilePublish(IEspContext &context, IEspDFUFilePublishRequest
+
         Owned<IDistributedFile> oldFile = queryDistributedFileDirectory().createNew(fileDesc);
         oldFile->validate();
 
-        if (!oldFile->renamePhysicalPartFiles(newFileName.str(), nullptr, nullptr, nullptr))
-            throw MakeStringException(ECLWATCH_FILE_NOT_EXIST, "Failed in renamePhysicalPartFiles %s.", newFileName.str());
+        CDfsLogicalFileName newLfn;
+        newLfn.set(newFileName);
+        if (!oldFile->renamePhysicalPartFiles(newLfn.queryTail(), nullptr, nullptr, fileDesc->queryDefaultDir()))
+            throw makeStringExceptionV(ECLWATCH_FILE_NOT_EXIST, "Failed in renamePhysicalPartFiles %s.", newFileName.str());
 
-        Owned<IFileDescriptor> newFileDesc = createFileDescriptor(newFileName, clusterTypeEx, groupName, group);
-        newFileDesc->queryProperties().setProp("@job", jobId);
-        if (!userId.isEmpty())
-            newFileDesc->queryProperties().setProp("@owner", userId);
+        StringBuffer newMask;
+        getPartMask(newMask, newFileName, fileDesc->numParts());
+        fileDesc->setPartMask(newMask);
 
-        Owned<IDistributedFile> newFile = queryDistributedFileDirectory().createNew(newFileDesc);
+        Owned<IDistributedFile> newFile = queryDistributedFileDirectory().createNew(fileDesc);
         newFile->setAccessed();
-        newFile->setECL(recordDefinition);
-        newFile->queryAttributes().setPropBin("_rtlType", layoutBin.length(), layoutBin.toByteArray());
+        newFile->attach(newFileName, userDesc);
+
+        LOG(daliAuditLogCat,",FileAccess,EspProcess,CREATED,%s,%s,%s", groupName, userId.str(), newFileName.str());
+    }
+    catch (IException *e)
+    {
+        exception.setown(e);
+    }
+
+    if (exception)
+    {
+        if (fileDesc)
+        {
+            Owned<IMultiException> exceptions = MakeMultiException("CWsDfuEx::onDFUFilePublish");
+            queryDistributedFileDirectory().removePhysicalPartFiles(normalizeTempFileName, fileDesc, exceptions);
+            if (exceptions->ordinality())
+            {
+                StringBuffer errMsg("Error whilst clearing up temporary file: ");
+                EXCLOG(exceptions, errMsg.append(normalizeTempFileName).str());
+            }
+        }
+        FORWARDEXCEPTION(context, exception.getClear(), ECLWATCH_INTERNAL_ERROR);
+    }
+    return true;
+}
+
+bool CWsDfuEx::onDFUPublishFile(IEspContext &context, IEspDFUPublishFileRequest &req, IEspDFUFilePublishResponse &resp)
+{
+    Owned<IException> exception;
+    Owned<IFileDescriptor> fileDesc;
+    StringBuffer normalizeTempFileName;
+    try
+    {
+        context.ensureFeatureAccess(FEATURE_URL, SecAccess_Write, ECLWATCH_DFU_ACCESS_DENIED, "WsDfu::DFUPublishFile: Permission denied.");
+
+        const char *fileId = req.getFileId();
+        if (isEmptyString(fileId))
+             throw makeStringException(ECLWATCH_INVALID_INPUT, "No FileId defined.");
+
+        const char *recordDefinition = req.getECLRecordDefinition();
+
```

Comment:
can we bail out of recordDefinition  is empty?

Reply:
It is quite normal to publish a file without a record definition. Sprayed files tend to be like that, so I don't want to make it illegal.